[dashboard] Allow Projects for Individuals

jankeromnes opened the pull request:
Fixes https://github.com/gitpod-io/gitpod/issues/4920

roboquat commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/4966#" title="Author self-approved">jankeromnes</a>*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [jankeromnes]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [jankeromnes]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [jankeromnes]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [jankeromnes]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

roboquat commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/4966#" title="Author self-approved">jankeromnes</a>*

*No associated issue*. Update pull-request body to add a reference to an issue, or get approval with `/approve no-issue`

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [jankeromnes]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [jankeromnes]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [jankeromnes]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [jankeromnes]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

roboquat commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/4966#" title="Author self-approved">jankeromnes</a>*

Associated issue: *#4920*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [jankeromnes]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [jankeromnes]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [jankeromnes]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [jankeromnes]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

jankeromnes commented on components/dashboard/src/projects/NewProject.tsx:
```diff
@@ -285,16 +287,17 @@ export default function NewProject() {
-                {teamsToRender.map((t) => (
-                    <div key={`team-${t.name}`} className={`w-96 border-b px-8 py-4 flex space-x-2 justify-between dark:hover:bg-gray-800 focus:bg-gitpod-kumquat-light transition ease-in-out group`}>
-                        <div className="w-8/12 m-auto overflow-ellipsis truncate">{t.name}</div>
+                {((user ? [user] : []) as Array<User | Team>).concat(teamsToRender).map((t) => {
```
Not super happy about this convoluted line :/ there is probably a better way.

roboquat commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **jankeromnes** after the PR has been reviewed.

Associated issue: *#4920*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**
- **[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)**
- **[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["jankeromnes"]} -->

jankeromnes commented:
Okay, I'd say this is (just barely) ready for review. 🙌

Will unmark draft, and prepare a deployment with a GitHub app, and then request reviews 🚀 (or, fix any last minute / obvious problems if things don't actually work 😅)

AlexTugarev commented on components/dashboard/src/projects/NewProject.tsx:
```diff
@@ -140,11 +140,13 @@ export default function NewProject() {
 
-        history.push(`/${team.slug}/${repo.name}/configure`);
+        history.push(User.is(teamOrUser)
+            ? '/projects' // FIXME
```
*Q*: shouldn't it be `/projects/repoName/configure`?

jankeromnes replied:
A: Yes, good catch!

jankeromnes replied:
@AlexTugarev So, like this?

- Individuals:
  - `/projects`
  - `/projects/:projectName/configure`
- Teams
  - `/:teamSlug/projects`
  - `/:teamSlug/:projectName/configure`

(I think this is slightly inconsistent in some sense, but doesn't shock me 🤷‍♂️)

AlexTugarev commented on components/dashboard/src/projects/Prebuild.tsx:
```diff
@@ -26,13 +26,20 @@ export default function () {
             return;
         }
         (async () => {
+            const projects = (!!team
```
this effect is bound to `[team]` and the early return is removed at L.29. 

I believe this might case a double call, if `setTeams` will return later due to slow network connection.

jankeromnes replied:
Ah, could be. Good catch, thanks!

jankeromnes replied:
Will add `!teams` to the early return `if`.

AlexTugarev commented on components/dashboard/src/projects/Prebuilds.tsx:
```diff
@@ -34,23 +34,19 @@ export default function () {
 
     useEffect(() => {
-        if (!team) {
-            return;
```
same here. 
I think this should be an early return if `teams` is not yet loaded, i.e. `setTeams` didn't return yet.

jankeromnes replied:
True. I agree we don't want the "teams are still loading..." state to temporarily show user projects. 🤦 Will adjust.

AlexTugarev commented:
`Projects.tsx` still contains at least one URL that is not aware of personal projects, i.e. points to `/${team?.slug}/...` and results in a broken link, e.g. `https://jx-user-projects.staging.gitpod-dev.com/undefined/bel`

AlexTugarev commented:
<img width="462" alt="Screen Shot 2021-08-02 at 15 42 08" src="https://user-images.githubusercontent.com/914497/127871288-f9b66a72-667e-48cb-ad1b-0079b984aad1.png">

I couldn't reach the project's page. The url always point to undefined. Also tried with `/projects/bel`, but this isn't found neither.

AlexTugarev commented:
When loading `/projects` server responds with `Error: Project not found`

![Uploading Screen Shot 2021-08-02 at 15.44.22.png…]()

AlexTugarev commented:
Navigating to `/projects` doesn't update `window.title`.

gtsiolis commented on components/dashboard/src/Menu.tsx:
```diff
@@ -80,6 +80,10 @@ export default function Menu() {
             return [
+                ...(showTeamsUI ? [{
+                    title: 'Projects',
+                    link: '/projects'
```
**question**: What do you think of defaulting to the first tab (projects) also for personal accounts? See also [relevant comment](https://github.com/gitpod-io/gitpod/issues/4920#issuecomment-886538402).

jankeromnes replied:
This is a bit tricky to achieve, because this changes the default tab for individuals based on whether the teams UI is enabled or not.

I'm not against such a change, but maybe it's safer to wait for the Teams & Projects UI to be generally enabled for everyone.

gtsiolis replied:
DEAL. Let's ship this and see what we can change later. ➿

gtsiolis replied:
FYI Added follow up issue for this in https://github.com/gitpod-io/gitpod/issues/5050. Cc @jankeromnes @AlexTugarev @svenefftinge

jankeromnes replied:
Thank you so much for not letting this slip by @gtsiolis! 💯 💯 💯

gtsiolis reviewed (commented):
Hey @jankeromnes! I was in the neighborhood and left a small comment below. 🚶

jankeromnes commented:
@AlexTugarev Okay, I believe I've now fixed all the Teams & Project URLs (now use either `/${team.slug}/` or `/projects/` as slug, depending on team- vs user-project).

I've also adjusted the top menu to work with individual projects as well, and I've made sure every page sets a title (by slightly abusing the `Header` component).

Please take another look / test again (deployment already has GH app) / let me know what you think! 🙏

AlexTugarev commented:
Comments/notes: 

* in general: having Projects under personal account is neat and it feels natural! 👍🏻 
* 🎉  thanks for fixing the `window.title` issue 💯 
* question: why is re-selecting of the personal account from the top menu navigating to `/workspaces` and not `/projects`? this is different with teams. still undecided, I'm not against it, just trying to understand the difference ;-)
* when opening the New Project wizard from the personal Projects page (New Project button below) it will ask me for the team. This is not the case when starting out from a team's Projects page. This seems to be inconsistent. Let's decide on a single strategy here. Maybe @gtsiolis can assist here? 
<img width="1377" alt="Screen Shot 2021-08-03 at 14 56 09" src="https://user-images.githubusercontent.com/914497/128020461-16e8bbcc-ebcd-402a-89b7-08ab0e6a5b6e.png">
* `Rerun Prebuild` seems to be coined to teams. At least it fails with `No team found`

gtsiolis commented:
> why is re-selecting of the personal account from the top menu navigating to `/workspaces` and not `/projects`?

@AlexTugarev I agree that eventually this could probably default to `/projects` just like with any other team. See also https://github.com/gitpod-io/gitpod/pull/4966#discussion_r681651234. It could be fine to also fix this in a smaller and separate PR if needed.

🍊 🍊 🍊 🍊 

> This is not the case when starting out from a team's Projects page. This seems to be inconsistent. Let's decide on a single strategy here. Maybe @gtsiolis can assist here?

When a user tries to create a project from within a team page it could make sense to automatically associate that project with the team they were looking at, right?

**suggestion**: When :a: going though the onboarding flow (`/new`) or :b: having the personal account selected, it could make sense for both cases to promt the user for a team selection including the personal account as a team option as seen in the relevant specs in https://github.com/gitpod-io/gitpod/issues/4948, see _Select Team_. Cross-posting below the relevant spec. Feel free to ignore the footnote about Teams & Projects pricing. What do you think? ❓ See also [relevant discussion](https://gitpod.slack.com/archives/C021Q4BL8E8/p1627987771044300?thread_ts=1627507356.092300&cid=C021Q4BL8E8).

| Team Selection for New Projects under Personal Account |
|-|
| <img width="1440" alt="127672643-cfc64766-0208-427a-8a9d-6c05b400dc74 (1)" src="https://user-images.githubusercontent.com/120486/128023436-270907d1-c67a-4ebb-8917-1eb57b3aa417.png"> |

svenefftinge commented:
> when opening the New Project wizard from the personal Projects page (New Project button below) it will ask me for the team. This is not the case when starting out from a team's Projects page. This seems to be inconsistent. Let's decide on a single strategy here.

I think it is good to ask in one but not the other case, because we want to help users put their projects below teams. So it makes sense to suggest it in that case. When I crate a project from a team there is no need to ask users to choose another team.

AlexTugarev commented:
After double-checking with gitpod-staging.com, it seems prebuilds aren't working here. I cannot see prebuilds running. Server logs do indicate prebuild registration, but there aren't any. @jankeromnes, could you check the webhook URL, please?

jankeromnes commented:
> After double-checking with gitpod-staging.com, it seems prebuilds aren't working here. I cannot see prebuilds running. Server logs do indicate prebuild registration, but there aren't any. @jankeromnes, could you check the webhook URL, please?

Interesting. With my integration, I was able to trigger a prebuild by:
1. Installing the app `janx-gitpod` on https://github.com/jankeromnes/gitpod-staging-prebuilds (while creating a new Project)
2. Creating a new Pull Request (i.e. branch) on that repository
3. See the new prebuild appear in the Prebuilds list

Not sure why you didn't see prebuilds. Maybe the `janx-gitpod` app wasn't installed on the repository you tried with? Or could be a different problem in this PR? (No idea where to look, but I could test again tomorrow.)

---

Also, what items are still blocking this PR from being merged?

- The "onboarding flow asks for which team when creating from user" inconsistency was brought up (because when creating from a team, the onboarding flow does not ask & pre-selects the relevant team). However, Sven pointed out that the inconsistency is desirable (asking again when creating from a user encourages people to create teams, which is what we prefer). So, no action required here.

- The "projects is the default tab for teams, but not for users" inconsistency was brought up. This should likely be fixed in the future (i.e. maybe Projects also the default tab for users), however doing that now is impractical, because of the Teams & Projects UI on/off flag (when off, we don't want a default page that "doesn't exist"). So, maybe we fix this later, when Teams & Projects is enabled generally.

- Prebuilds didn't work for Alex. Jan can try to test prebuilds again tomorrow.

- ... anything else still blocking a lgtm? 😇

jankeromnes commented:
Gitpod deployment, I summon thee! 🧙

/werft run

svenefftinge commented:
> anything else still blocking a lgtm? 😇

FWIW, to me it lVerygtm 🎉 when I tested it yesterday and looked over the code, but since you had intense conversations with Alex, I think he should do the approval.

AlexTugarev commented:
> * `Rerun Prebuild` seems to be coined to teams. At least it fails with `No team found`

@jankeromnes, this still fails.

AlexTugarev commented:
<img width="1434" alt="Screen Shot 2021-08-04 at 08 55 55" src="https://user-images.githubusercontent.com/914497/128135783-95dfe6e6-5945-4138-857c-86a2182b49bd.png">

🎉  Prebuilds are shown in personal and team project!

jankeromnes commented:
Woohoo! Thanks Alex! 🎉

(To be safe, I even recorded a quick [screencast](https://youtu.be/0ttNZvt17Yk) to demo prebuilds in action 😁 but I guess it's no longer needed!)

> * `Rerun Prebuild` seems to be coined to teams. At least it fails with `No team found`

Oh! I totally overlooked this comment. 🙈 Will check & fix.

AlexTugarev commented on components/dashboard/src/projects/ConfigureProject.tsx:
```diff
@@ -62,14 +62,13 @@ export default function () {
   const [ workspaceCreationResult, setWorkspaceCreationResult ] = useState<WorkspaceCreationResult | undefined>();
 
   useEffect(() => {
-    if (!team) {
```
@jankeromnes should we add a check for "`teams` loaded" here as well?

jankeromnes replied:
Uh... yes. 😳 I thought I got them all. Probably need more ☕!

AlexTugarev commented:
/lgtm 👍🏻 👍🏻 👍🏻 

/hold @jankeromnes, please have a look at the last two comments of mine.

roboquat commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/4966#issuecomment-892418594" title="LGTM">AlexTugarev</a>*

Associated issue: *#4920*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [AlexTugarev]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [AlexTugarev]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [AlexTugarev]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [AlexTugarev]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

jankeromnes commented:
Fixed the two nits, and added a dependency on `teams` for all "project loading effects": https://github.com/gitpod-io/gitpod/compare/df93235422df3d3646f4e7f565c3d644eaa4d11a..98080ef4db388dfc0acec48e779ba416894f7bf8

Carrying over the /lgtm from https://github.com/gitpod-io/gitpod/pull/4966#issuecomment-892418594 😇
